src/include/platform/internal/GenericPlatformManagerImpl.ipp

```diff
@@ -79,9 +79,47 @@ CHIP_ERROR GenericPlatformManagerImpl<ImplClass>::_InitChipStack(void)
     }
     SuccessOrExit(err);
 
-    // TODO Perform dynamic configuration of the core CHIP objects based on stored settings.
+#if CHIP_ENABLE_MESSAGE_LAYER
+    // Initialize the CHIP Fabric State object.
+    new (&FabricState) ChipFabricState();
```

Comment:
What's this for? There's already an object at |FabricState|.

Reply:
This basically calls the constructor of class ChipFabricState using the existing FabricState object instead of allocating new memory from the heap.

This pattern is used to initialize all CHIP core components, such as:
new (&SystemLayer) System::Layer();
new (&InetLayer) Inet::InetLayer();

Yes, the constructor of those objects has already been called by compiler during the object initialization, but that initialization is invalid since those components have dependency, we need to re-initialize them sequentially with specified order.

Reply:
No, all core components are not created with new operator, they are global static objects.

chip::System::Layer SystemLayer;
chip::Inet::InetLayer InetLayer;

Reply:
This is overwriting a constructed object. My request is to avoid doing that without doing one of the following:

- calling the destructor
- ensuring (e.g. using static_assert) that the type is trivially destructible:

```
   static_assert(std::is_trivially_destructible<ChipFabricState>::value,
                         "ChipFabricState must be trivially destructable");
```

If neither assumption holds we're risking undefined behavior.

However, it may be even better to just move these objects into a static function and avoid the placement new altogether:

```
FabricState& GetFabricState() {
  static FabricState sFabricState;
  return sFabricState;
}
```
This avoids the initialization order problem.

Reply:
Do you know what the source of the init order is? It's not really clear what fields are order dependent.

Reply:
`FabricState` is actually not trivially destructible, because it has a field of type `MonotonicallyIncreasingCounter` which has a virtual destructor.

Reply:
I agree using new operator to init but not allocate memory for a object is really confusing. 

I think we should also avoid using other advanced techniques to work around this issue. I moved all functional initialization from constructor to init function, then we can avoid using new to re-init those objects.

Reply:
While a somewhat awkward construct, the in-place new operator allows us to use static memory allocation.  For embedded devices, this is important, as it provides deterministic memory usage of an application which is immediately visible using `arm-none-eabi-size` of the image.  Ideally we would eliminate any dynamic allocation from all the CHIP examples, especially for embedded platforms.

Reply:
> Ideally we would eliminate any dynamic allocation from all the CHIP examples, especially for embedded platforms.

Not just ideally; it's a requirement.

Reply:
This looks good now. It might be a good idea to do this broadly, after which we could `static_assert(std::is_pod<T>::value)` on the storage type to ensure no static initializers creep back in.

As a point of reference objects of static storage duration that aren't trivially destructible are banned by the Google C++ style guide.

Reply:
@turon, per my understanding, "new (&FabricState) ChipFabricState();" does not allocate memory, it just calls the constructor of class ChipFabricState using the existing FabricState object. 

If we move all functional initialization to init function from constructor, then we don't need to call this in-place new operator during platform initialization. Did I miss anything?

Reply:
You don't need to call placement new to construct an object of static storage duration. That happens automatically. If we make a duplicate constructor call, we may cause undefined behavior in trying to construct an object over another object whose lifetime has not ended. That's what I was objecting to. Use of static or dynamic memory allocation is orthogonal to this.

Reply:
> You don't need to call placement new to construct an object of static storage duration. That happens automatically. If we make a duplicate constructor call, we may cause undefined behavior in trying to construct an object over another object whose lifetime has not ended. That's what I was objecting to. Use of static or dynamic memory allocation is orthogonal to this.

I think this misunderstands the idiom here that Weave has historically used. Static *objects* were never allocated or constructed since such construction often requires a primoridal heap which didn't / doesn't exist on such systems and required locking constructs that not only didn't exist but were expensive to implement.

So, the idiom was to allocate a buffer of bytes for the object and then to do placement new on that buffer followed by an explicit initializer:

```
DEFINE_ALIGNED_VAR(sFoo, sizeof(Foo), void *); // At global scope,  no primordial object creation

{
    Error lRetval;

    sFooPointer = new (sFoo) Foo();
    VerifyOrExit(sFooPointer != NULL);

    lRetval = sFooPointer->Init();
    SuccessOrExit(lStatus);

    exit:
        return (lRetval);
}
```

Reply:
static objects not being constructed is non-conformant behavior in the C++ language runtime. I agree that such behavior will be more common in an embedded context (where I've encountered even more serious bugs such as unzeroed bss).

The answer is to use trivial types, which avoids generating static initializers in the first place. Using a std::aligned_storage or equivalent accomplishes exactly that. If that's what weave does (or used to do ?), it helps explain how we ended up in this broken state of constructing everything twice. It doesn't however change my objection, which is that the code as it exists in CHIP is dangerous and should be fixed.

Reply:
@mspang, 
The current revision should be able to address our concerns, even the static object are not properly initialized, but the constructor of those static objects are always called by  C++ standard. Now the constructor of those objects only contains default value initialization, no sequence dependency on each other. So we can skip the in-place new on those objects, and only call the explicit initializer of those objects in correct sequence.